src/libraries/System.Private.CoreLib/src/System/IO/Strategies/AsyncWindowsFileStreamStrategy.ValueTaskSource.cs

```diff
@@ -0,0 +1,248 @@
+                // Only one operation at a time is eligible to use the preallocated overlapped
+                _strategy.CompareExchangeCurrentOverlappedOwner(null, this);
+            }
+
+            private static void IOCallback(uint errorCode, uint numBytes, NativeOverlapped* pOverlapped)
+            {
+                // Extract the AwaitableProvider from the overlapped.  The state in the overlapped
+                // will either be a AsyncWindowsFileStreamStrategy (in the case where the preallocated overlapped was used),
+                // in which case the operation being completed is its _currentOverlappedOwner, or it'll
+                // be directly the AwaitableProvider that's completing (in the case where the preallocated
+                // overlapped was already in use by another operation).
+                object? state = ThreadPoolBoundHandle.GetNativeOverlappedState(pOverlapped);
+                Debug.Assert(state is AsyncWindowsFileStreamStrategy or ValueTaskSource);
+                ValueTaskSource valueTaskSource = state switch
+                {
+                    AsyncWindowsFileStreamStrategy strategy => strategy._currentOverlappedOwner!, // must be owned
+                    _ => (ValueTaskSource)state
+                };
+                Debug.Assert(valueTaskSource != null);
+                Debug.Assert(valueTaskSource._overlapped == pOverlapped, "Overlaps don't match");
+
+                // Handle reading from & writing to closed pipes.  While I'm not sure
+                // this is entirely necessary anymore, maybe it's possible for
+                // an async read on a pipe to be issued and then the pipe is closed,
+                // returning this error.  This may very well be necessary.
+                ulong packedResult;
+                if (errorCode != 0 && errorCode != Interop.Errors.ERROR_BROKEN_PIPE && errorCode != Interop.Errors.ERROR_NO_DATA)
+                {
+                    packedResult = ((ulong)TaskSourceCodes.ResultError | errorCode);
+                }
+                else
+                {
+                    packedResult = ((ulong)TaskSourceCodes.ResultSuccess | numBytes);
+                }
+
+                // Stow the result so that other threads can observe it
+                // And, if no other thread is registering cancellation, continue
+                if (Interlocked.Exchange(ref valueTaskSource._result, (long)packedResult) == TaskSourceCodes.NoResult)
+                {
+                    // Successfully set the state, attempt to take back the callback
+                    if (Interlocked.Exchange(ref valueTaskSource._result, TaskSourceCodes.CompletedCallback) != TaskSourceCodes.NoResult)
+                    {
+                        // Successfully got the callback, finish the callback
+                        valueTaskSource.CompleteCallback(packedResult);
+                    }
```

Comment:
I know you're copying this logic from what was there before, but it seems to be unnecessarily complex, with lots of interlocked operations for transitioning from one state to another to another.  If we're looking for places to reduce overheads, revisiting this whole implementation is likely a good place to start.